### ISSUE TITLE
Rules that define a valid curated cell data file

### ISSUE THREAD

**Issue by philerooski** [Sage-Bionetworks/neurolincsdreamchallenge#17]
I will be automatically checking that each track in the curated cell data file satisfies some constraints. Here's a list of those constraints (@jlinsley1983 verify these rules make sense or if you can think of any I may be missing). The idea is that if one or more of these constraints is not met, we have one or more problem tracks that need fixing.

**General**
1. All TimePoints must be present w.r.t. the Experiment (no missing TimePoint values).

**Live_Cells**
1. If Live_Cells is FALSE, all subsequent Live_Cells values must be FALSE.

**Lost_Tracking**
1. If Lost_Tracking is TRUE, all subsequent Lost_Tracking values must be TRUE.
2. If Lost_Tracking is TRUE, ObjectLabelsFound must be NA

**Comment by kdaily** [comment]
Needs to be a timepoint 0.

**Comment by philerooski** [comment]
That's covered by
> All TimePoints must be present w.r.t. the Experiment (no missing TimePoint values).

I use this table as a reference to determine which TimePoint values must be present: https://www.synapse.org/#!Synapse:syn11817859/tables/

**Comment by jlinsley1983** [comment]
I'm not sure I understand your last comment on why that is covered by no missing timepoint values. To be clear, there are some tracks in the data that start on time point 4 for example, and do not have any missing timepoint values. They should not be counted here though, as we are only scoring tracks that start at timepoint 0 and do not have any missing timepoints.

**Comment by philerooski** [comment]
That's right. These rules only apply to the tracks we are using for scoring. So tracks that come from an Experiment that start at TimePoint = 4 have already been removed.

**Comment by philerooski** [comment]
@jlinsley1983 Is this true?

> If Lost_Tracking is TRUE, all subsequent Lost_Tracking values must be TRUE.

I see quite a few tracks like this:
![image](https://user-images.githubusercontent.com/2773709/73093569-a5bd9080-3e93-11ea-8a9a-f12c4dd30849.png)

Which looks like tracking was lost, then found again at a later timepoint, before being lost again until the end of the experiment.

**Comment by jlinsley1983** [comment]
Yes. In this particular case, there is no segmentation mask for #16 at timepoint 5, even though a cell is still actually present in the image. The object came back at a later timepoint, so the curator indicated that object is still alive at later timepoints (as different object numbers). Once Lost_tracking=TRUE in a track, it should stay TRUE.

**Comment by philerooski** [comment]
Does that fact affect the validity of this rule?
> If Lost_Tracking is TRUE, ObjectLabelsFound must be NA

If the segmentation mask appears again at TimePoint 8 as it did in the previous example, it's plausible that we should retain the ObjectLabelsFound value at that TimePoint rather than set it to NA.

**Comment by jlinsley1983** [comment]
Hi Phil,

The issue is that if there is no segmentation mask, there is no way that the tracking algorithms can get a track correct. Asking an algorithm to skip several timepoints and still track an object is unrealistic, and will likely lead to a worse results in an algorithm if we ask it to perform that task. It should still be a minority of cases in which a track is lost, and it will be easy to censor lost tracks from our data.

I think if Lost_Tracking is TRUE, all ObjectLabelsFound after that point should be changed to NA. There is no benefit to tracking objects after a segmentation mask has been lost, just as we decided within our framework that there is no additional benefit to tracking new objects that come up after timepoint 1. 


Best,
Jeremy

**Comment by philerooski** [comment]
Thanks! Formalizing these rules helps a lot.

Another one -- I found this comment in my code but I can't find a source for it among the preexisting issues.
> If Lost_Tracking = TRUE, we don't know Live_Cells value, so set Live_Cells to NA

Is this true?

**Comment by jlinsley1983** [comment]
In principle I agree, but can you send me a specific example (s) to check?

Thanks,
Jeremy

**Comment by philerooski** [comment]
For example in this track:

```
# A tibble: 17 x 7
   Experiment           Well  ObjectTrackID ObjectLabelsFound TimePoint Live_Cells Lost_Tracking
   <chr>                <chr>         <int>             <int>     <int> <lgl>      <lgl>        
 1 AB-CS47iTDP-Survival A1               16                16         0 TRUE       FALSE        
 2 AB-CS47iTDP-Survival A1               16                16         1 TRUE       FALSE        
 3 AB-CS47iTDP-Survival A1               16                63         2 TRUE       FALSE        
 4 AB-CS47iTDP-Survival A1               16                75         3 TRUE       FALSE        
 5 AB-CS47iTDP-Survival A1               16                98         4 TRUE       TRUE         
 6 AB-CS47iTDP-Survival A1               16               192         9 TRUE       FALSE        
 7 AB-CS47iTDP-Survival A1               16               209        10 TRUE       FALSE        
 8 AB-CS47iTDP-Survival A1               16               225        11 TRUE       FALSE        
 9 AB-CS47iTDP-Survival A1               16               225        12 TRUE       FALSE        
10 AB-CS47iTDP-Survival A1               16               225        13 TRUE       FALSE        
11 AB-CS47iTDP-Survival A1               16               262        14 TRUE       FALSE        
12 AB-CS47iTDP-Survival A1               16               272        15 TRUE       FALSE        
13 AB-CS47iTDP-Survival A1               16               285        16 TRUE       FALSE        
14 AB-CS47iTDP-Survival A1               16               294        17 TRUE       FALSE        
15 AB-CS47iTDP-Survival A1               16               294        18 TRUE       FALSE        
16 AB-CS47iTDP-Survival A1               16               309        19 TRUE       FALSE        
17 AB-CS47iTDP-Survival A1               16               309        20 TRUE       TRUE 
```
We lost tracking past TimePoint = 4, but we continue to record Live_Cells = TRUE. There are other tracks where we have lost track of the cell (Lost_Tracking = TRUE) but we continue to record Live_Cells = FALSE at later TimePoints.

**Comment by jlinsley1983** [comment]
Yes, looks good. In this case, the object 98 track is lost (because the mask disappears), though the cell does not appear to die at that time point. Strangely enough, the curator indicates that it changes to #192 and then #209, 225 etc even though that doesn’t appear to be the case. So it is definitely appropriate to change all Live_Cells to NA after the initial Lost_tracking=TRUE.

Best,
Jeremy
